accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1008,6 +1011,33 @@ void main() {
 }
 ```
 
+### Constant assertions
+
+Currently assertions are enabled only when the compiler is passed an `enable-asserts` flag.
```

Comment:
The text in section 'Assert' of the language specification already says almost everything which is said here.

Wouldn't it suffice to have a very short section here? We would then rely on the existing specification text for all the details:

```suggestion
Assertions are enabled during the evaluation of a constant object expression.

*This means that properties checked by assertions will always be enforced with constant expressions. The implementation specific mechanism to enable assertions is used to enable the remaining assertions, namely the ones that are executed at run time.*
```

Reply:
It's tricky, because we don't actually separate compile-time and run-time in the specification, we just specify the meaning of constants such that they *can* be pre-computed, but we don't actually say that you have to do so.

We still need a way to specify that something is happening as part of constant evaluation because it matters in some situations. For example `const x = 1 ~/ 0;` is a compile-time error instead of throwing a `DivisionByNullError`.

Reply:
Right, we specify that evaluation of a constant object expression occurs at run time and then it is (recursively, bottom-up) canonicalized. It makes no difference whether that constant object expression is in a constant context, whether it is used in the initialization of a `const` variable, etc., it is a rule for `<constObjectExpression>` anywhere. Of course, if it is in a constant context then the `const` keyword may be absent, it is still statically known to be there.

This rule is compatible with an evaluation at compile time. However, it doesn't _require_ an implementation to perform the evaluation at compile time and reuse that result every time it is needed at run time. For instance, the expression could be evaluated at compile time and the result could be discarded, and then evaluation and canonicalization could be done at run time. Also, because the constant sublanguage is small and Turing-incomplete, execution is guaranteed to be possible at compile time. So when we want to determine whether an error should be emitted, we can evaluate the constant object expression.

So if we do say 'Assertions are enabled during the evaluation of a constant object expression' then it is allowed for an implementation to raise the compile-time error early ("at compile time"). That's probably exactly what we want.

This also means that _if_ a tool gives rise to an execution where a constant object expression is actually evaluated at run time then there will be a run-time error, even though we set out to say that these errors should occur at compile-time. But tools don't have to do this, so in practice we can maintain that these errors occur early.

If an execution uses hot reload then compilation will occur at run time anyway.

Don't you think that would be OK for tool teams to work with?